src/main/relation/join_relation.cpp

```diff
@@ -26,9 +30,55 @@ JoinRelation::JoinRelation(shared_ptr<Relation> left_p, shared_ptr<Relation> rig
 	context.GetContext()->TryBindRelation(*this, this->columns);
 }
 
+JoinRelation::JoinRelation(shared_ptr<Relation> left_p, shared_ptr<Relation> left_expr, shared_ptr<Relation> right_proj,
+                           JoinType type)
+    : Relation(left_p->context, RelationType::JOIN_RELATION), left(move(left_p)), right(move(right_proj)),
+      left_expr(left_expr), using_columns(), join_type(type) {
+	if (join_type != JoinType::ANTI && join_type != JoinType::SEMI) {
+		throw Exception("Must pass conditions for join of type " + JoinTypeToString(join_type));
+	}
+	if (left->context.GetContext() != right->context.GetContext()) {
+		throw Exception("Cannot combine LEFT and RIGHT relations of different connections!");
+	}
+	if (right->type != RelationType::PROJECTION_RELATION) {
+		throw Exception(JoinTypeToString(join_type) + " requires a projection for the right relation. Received " +
+		                RelationTypeToString(right->type));
+	}
+	condition = nullptr;
+	context.GetContext()->TryBindRelation(*this, this->columns);
+}
+
 unique_ptr<QueryNode> JoinRelation::GetQueryNode() {
 	auto result = make_unique<SelectNode>();
 	result->select_list.push_back(make_unique<StarExpression>());
+	if (join_type == JoinType::ANTI || join_type == JoinType::SEMI) {
```

Comment:
I am in favor of it being a separate function.  Especially since you pointed out that the join is considered a "filter" join by dplyr. Would make the join logic as a whole readable

Reply:
If they were, merging `master` into this branch would clean up the diff.